Add OpenTelemetry tracing to some RunEngine methods

dperl-dls opened the pull request:
## Description

Submitted together with https://github.com/bluesky/ophyd/pull/1189

- adds `opentelemetry-api` as a dependency
- wraps some potentially long-running `RunEngine` methods (`wait()`,`set()`,`complete()`...) in tracing spans
- creates spans at `_open_run` which are ended on `_close_run` (or abort, halt etc)
- attaches some metadata about each of the above

## Motivation and Context
Fixes #1682 

## How Has This Been Tested?
The `api` component of `OpenTelemetry` is designed to not really do anything on its own - I welcome suggestions about tests to include here.
This was tested by setting up a `TracingProvider` in Hyperion and exporting the results from running our unit tests to a Jaeger container.

Making use of these features in Hyperion is currently in [this branch](https://github.com/DiamondLightSource/hyperion/tree/bluesky_1682_improve_tracing), and an example of how to set up a tracing provider can be found [here](https://github.com/DiamondLightSource/hyperion/blob/bluesky_1682_improve_tracing/src/hyperion/tracing.py).



![screenshot-2024-04-17-16:04:56](https://github.com/bluesky/bluesky/assets/115003895/87bb1212-534d-4ae3-954c-4b2840171401)

![screenshot-2024-04-17-16:06:23](https://github.com/bluesky/bluesky/assets/115003895/a00e49ec-a243-41f8-af99-6b07add4754c)

![screenshot-2024-04-17-16:30:56](https://github.com/bluesky/bluesky/assets/115003895/a5f00a22-3b05-4397-bd1e-311f368edd4a)

callumforrester commented:
Power cycle

mrakitin commented on src/bluesky/tests/test_tracing.py:
```diff
@@ -0,0 +1,26 @@
+        yield from sleep(0)
+        assert not get_current_span().is_recording()
+
+    RE = RunEngine()
```
Why not to use the `RE` fixture?

mrakitin reviewed (commented):
Looks good overall. I have a couple of questions here:
1) How easy it will be to opt-out from tracing?
2) I wonder if we could apply the traces as a `RE.msg_hook`?

dperl-dls commented:
Thanks for the review @mrakitin !
1. Opting out from tracing means just doing nothing, to actually do anything with these tracing spans is opt-in. Each of these tracing spans is essentially a no-op (well, an `if false`), unless some application code has used the `opentelemetry-sdk` library to set up some tracing provider and exporter. They're basically just hooks for application tracing code to inject into. If you run some code as in https://github.com/DiamondLightSource/hyperion/blob/bluesky_1682_improve_tracing/src/hyperion/tracing.py to set up exporters, then those hooks will be used to determine what info to export.
2. I will look in to it - my immediate guess is that it might be difficult because these traces are meant to wrap functions or be created and destroyed according to some logical unit of work the program is doing, which means we need some corresponding hook at the end of message processing. It's a little more complex than I really wanted this PR to be, but if there is interest I could look at making a `msg_context_hook` or something like that to wrap message processing in some given context manager, which would allow that to work and would be quite elegant IMO.

danielballan commented:
I am not so sure this is a good fit for `msg_hook`. The `msg_hook` receives a generic input (`Msg`) and performs a generic action (log a line of text). Telemetry Traces are richer things: each _type_ of Msg (each `msg.command`) may suggest custom `attributes` and so on that are best handled inside the specific coroutine where the relevant context is held.

---

1. Should Msg types be differentiated by a part of a span name that has to be parsed out of a string (`Bluesky RunEngine set`, as currently implemented) or via an `attribute` (`set`)? Or both?
2. Should the message content by logged as a single attribute or as separate attributes for each piece (`msg.command`, `msg.args`, `msg.kwargs`, `msg.obj`)?
3. I notice the use of `repr` for generating strings to place in the spans. In the past I've regretting using `repr` for this when the string may be parsed by code in other languages. A more strictly defined serialization (e.g. JSON) can be better.
4. Could a documentation page be added describing the schemas of the events emitted? This could serve as a both a reference for a developers and a helpful design document to aid discussions about how to improve/extend our use of OpenTelemetry.

dperl-dls commented:
Hi @danielballan , sorry I left this for so long - I've added a bunch of finer grained attributes, which I agree should be much more useful. I left in the `repr`s because they are pretty handy human readable things when looking through the traces by hand. I've also added some docs, mainly with a view of being a reference for devs.

danielballan reviewed (approved):
This addresses all my questions and suggestions. Seems solid. Thanks for including docs.
